Minimal scripting support

philwalk opened the pull request:
This supercedes #11180

Implements scala3 scripting support with the minimum functionality required for supporting real world scripting:

- should be familiar to scala2 script developers, following the same basic design.
- hash-bang section is ignored by the compiler
- error message line numbers are correct even with hash-bang section
- no constraint on script name, if hash-bang section is present (extension allowed by not required)
- script runs in compiler process, avoiding multiple `jvm` startups
- on successful compile, `ScriptingDriver` calls client callback prior to calling script `main` method.
- The `-save` command line option writes a jar file to the script parent directory (same as scala2)
- jar file manifest is written with `Main-Class` and `Class-Path` properties (jar is executable)

The client callback is needed to implement the `-save` option with a single `jvm` startup.

3rd party tools are able to implement custom implementations following the example in `dotty.tools.scripting.Main`.

SethTisue commented:
👏   I am eager to test this once it lands in a nightly

philwalk commented:
resolves these issues:
#10491
#4747
#10858

anatoliykmetyuk commented on dist/bin/scala:
```diff
@@ -82,9 +92,12 @@ while [[ $# -gt 0 ]]; do
     *)
       if [ $execute_script == false ]; then
-        if [[ "$1" == *.scala ]]; then
+        # is a script if extension .scala or .sc or if has scala hash bang
```
This conflicts with the definition of `isScript` in `SourceFile` which says that a script must not have an extension `*.scala`.

philwalk replied:
good catch, that was intended for a possibly slight performance gain, but might not have actually provided one anyway.

anatoliykmetyuk commented on dist/bin/scala:
```diff
@@ -82,9 +92,12 @@ while [[ $# -gt 0 ]]; do
           execute_script=true
+          [ -n "$SCALA_OPTS" ] && java_options+=($SCALA_OPTS)
           target_script="$1"
+          if [ ! -f $target_script
```
Missed closing `]` here.

anatoliykmetyuk replied:
I believe that’s why the CI test failed. Does the `scala` script work locally?

philwalk replied:
I fixed this and pushed the change 30 min. ago, and will push your requested changes within the hour

anatoliykmetyuk commented on dist/bin/scala:
```diff
@@ -101,7 +114,17 @@ if [ $execute_script == true ]; then
+  if [[ $jar_found == true && "$target_jar" -nt "$target_script" ]]; then
+    java $setScriptName -jar "$target_jar" "${script_args[@]}"
+  else
+    [[ $save_compiled == true && -f $target_jar ]] && rm -f $target_jar
```
You could replace the first operand of `&&` with `$jar_found == true`.

philwalk replied:
I reworked that section, removing `$jar_found`:
```
  if [[ $save_compiled == true && "$target_jar" -nt "$target_script" ]]; then
```

anatoliykmetyuk commented on compiler/src/dotty/tools/dotc/util/SourceFile.scala:
```diff
@@ -22,19 +22,35 @@ object ScriptSourceFile {
       else 0
-    new SourceFile(file, content drop headerLength) {
+
+    // overwrite hash-bang lines with all spaces
```
This is done to preserve the positions, right? If so, we should say so explicitly in the comment.

philwalk replied:
done, with push after tests run

anatoliykmetyuk reviewed (changes requested):
Looks good to me, nice work! After the minor suggestions are addressed, this can be merged.

philwalk commented on compiler/src/dotty/tools/scripting/Main.scala:
```diff
@@ -1,22 +1,82 @@
+    import dotty.tools.io.{Jar, Directory}
+    val jar = new Jar(jarPath)
+    val writer = jar.jarWriter(manifestAttributes:_*)
+    writer.writeAllFrom(Directory(outDir))
```
done, will check code within the hour

philwalk commented on compiler/src/dotty/tools/dotc/util/SourceFile.scala:
```diff
@@ -245,6 +261,24 @@ object SourceFile {
+    if isScript(file, chars) then
+      ScriptSourceFile(file, chars)
+    else
+      new SourceFile(file, chars)
```
this seems to be necessary to distinguish between constructor and apply method

philwalk commented on compiler/src/dotty/tools/dotc/util/SourceFile.scala:
```diff
@@ -245,6 +261,24 @@ object SourceFile {
+    if file.hasExtension(".scala") then
+      false
+    else
+      ScriptSourceFile.hasScriptHeader(content)
```
ah, much better.

philwalk replied:
I removed `!file.hasExtension(".scala")` so a hash-bang section will be ignored in all scala source files.

philwalk commented:
@anatoliykmetyuk - am I supposed to `[Resolve conversation]` or do you do that?

anatoliykmetyuk commented:
> @anatoliykmetyuk - am I supposed to `[Resolve conversation]` or do you do that?

Feel free to resolve them if no further input is needed!

philwalk commented on compiler/src/dotty/tools/scripting/Main.scala:
```diff
@@ -1,22 +1,82 @@
+  private def distinguishArgs(args: Array[String]): (Array[String], File, Array[String], Boolean) =
+    val (leftArgs, rest) = args.splitAt(args.indexOf("-script"))
+    if( rest.size < 2 ) then
+      sys.error(s"missing: -script <scriptName>")
```
We should emit an error message, so maybe it should say:
`internal error: -script <scriptName> missing`

philwalk commented on compiler/src/dotty/tools/scripting/Main.scala:
```diff
@@ -1,22 +1,82 @@
+    val cpPaths = runtimeClasspath.split(pathsep).map {
+      // protect relative paths from being converted to absolute
+      case str if str.startsWith(".") && File(str).isDirectory => s"${str.withSlash}/"
+      case str if str.startsWith(".") => str.withSlash
```
These cases deal with `classpath` entries for directory and jar relative paths.
Without this, they are converted to absolute paths here:

`File(str).toURI.toURL.toString`

Using forward slash is for simplicity and consistency, since URLs have forward slashes (even on Windows).

philwalk commented:
@anatoliykmetyuk - I have some refinements that I can push, but I thought I should check with you first.   Here's what I have:

1. I realized that 3rd party tools would likely have a use case for compiling without execution, so I added a "-compile-only" option
2. after lots of manual testing, I refined the detection of manifest class-path relative path entries
3. I added new tests to verify the new "-compile-only" option, plus some other tests and updated comments

Let me know if it's okay to push these changes, I don't want to risk delaying the schedule.

anatoliykmetyuk reviewed (approved):
LGTM, thanks @philwalk!
